### PR TITLE
[AP-953] Add parquet support

### PR DESCRIPTION
**Pull request by koszti**
## Problem

Target-snowflake is using CSV files to load data from s3 into tables, but ~CSV is not the most efficient file format when loading into columnar tables.~

## Proposed changes

Add optional parquet support. To select between CSV and Parquet file format, create the file format snowflake object by one of the following SQLs:

To use CSV:
```
CREATE FILE FORMAT {database}.{schema}.{file_format_name}
TYPE = 'CSV' ESCAPE='\\' FIELD_OPTIONALLY_ENCLOSED_BY='"';
```

To use Parquet:
```
CREATE FILE FORMAT {database}.{schema}.{file_format_name} TYPE = 'PARQUET';
```

The required file format specific functions will be selected automatically. The code detects the file format type by running `SHOW FILE FORMATS LIKE <file_format_name>` SQL command

**Key functions in the code:**
* [parquet.py: records_to_dataframe](https://github.com/transferwise/pipelinewise-target-snowflake/pull/149/files#diff-8982d5918e61681863141ec5189a1953cc27b988e5b1ed2f286a6b5446733627R55): Transforms a batch of singer record messages to pandas dataframe
* [parquet.py: dataframe.to_parquet](https://github.com/transferwise/pipelinewise-target-snowflake/pull/149/files#diff-8982d5918e61681863141ec5189a1953cc27b988e5b1ed2f286a6b5446733627R112): Writes the DataFrame to the binary parquet format.
* [parquet.py: create_merge_sql](https://github.com/transferwise/pipelinewise-target-snowflake/pull/149/files#diff-8982d5918e61681863141ec5189a1953cc27b988e5b1ed2f286a6b5446733627R28): Snowflake MERGE SQL generator for Parquet files
* [csv.py: create_merge_sql](https://github.com/transferwise/pipelinewise-target-snowflake/pull/149/files#diff-ba53b5ad6056c1a6a140cc82ac05cf3bf167cda803fca1920e46be4bb7deb1deR28): Snowflake MERGE SQL generator for CSV files.

## Types of changes

What types of changes does your code introduce to PipelineWise?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

- [x] Description above provides context of the change
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Unit tests for changes (not needed for documentation changes)
- [x] CI checks pass with my changes
- [x] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [x] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [x] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: target_snowflake/upload_clients/s3_upload_client.py [hunk 1]
```diff
@@ -50,9 +50,10 @@ def upload_file(self, file, stream, temp_dir=None):
         bucket = self.connection_config['s3_bucket']
         s3_acl = self.connection_config.get('s3_acl')
         s3_key_prefix = self.connection_config.get('s3_key_prefix', '')
-        s3_key = "{}pipelinewise_{}_{}.csv".format(s3_key_prefix,
-                                                   stream,
-                                                   datetime.datetime.now().strftime("%Y%m%d-%H%M%S-%f"))
+        s3_key = "{}pipelinewise_{}_{}_{}".format(s3_key_prefix,
```
Comment (Samira-El) [inline target_snowflake/upload_clients/s3_upload_client.py:53]: Can you share an example of how this s3_key will look like with change?

File: target_snowflake/file_formats/parquet.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                    file_format_name: str,
+                    columns: List):
+    """Generate a Parquet compatible snowflake COPY INTO command"""
+    return "COPY INTO {} ({}) " \
+           "FROM (SELECT {} FROM '@{}/{}') " \
+           "FILE_FORMAT = (format_name='{}')".format(
+        table_name,
+        ', '.join([c['name'] for c in columns]),
+        ', '.join(["{}($1:{}) {}".format(c['trans'],
+                                         c['json_element_name'],
```
Comment (Samira-El) [inline target_snowflake/file_formats/parquet.py:23]: Can you elaborate a bit on why is this `json_element_name` needed and used for?

File: target_snowflake/db_sync.py [hunk 3]
```diff
@@ -204,6 +207,11 @@ def __init__(self, connection_config, stream_schema_message=None, table_cache=No
         self.grantees = None
         self.file_format = FileFormat(self.connection_config['file_format'], self.query)
 
+        if not self.connection_config.get('stage') and self.file_format.file_format_type == FileFormatTypes.PARQUET:
+            self.logger.error("Table stages with Parquet file format is not suppported. "
+                              "Use named stages with Parquet file format or table stages with CSV files format")
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (koszti) [reply target_snowflake/db_sync.py:213]: just to keep it consistent with other exit few lines above. We can address this separately by introducing a new `ConfigurationValidationException` ?

File: target_snowflake/file_format.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+        Params:
+            file_format: File format name
+            query_fn: A callable function that can run SQL queries in an active Snowflake session
+
+        Returns:
+            FileFormatTypes enum item
+        """
+        file_format_name = file_format.split('.')[-1]
+        file_formats_in_sf = query_fn(f"SHOW FILE FORMATS LIKE '{file_format_name}'")
```
[the remark this answers is not in the document]
Comment (koszti) [reply target_snowflake/file_format.py:52]: `DESC` and `SHOW FILE FORMATS LIKE 'exact_name'` gives logically equivalent results, but parsing the output of SHOW was easier. `DESC` returns file format type in rows while `SHOW` returns it in a dedicated column.

File: target_snowflake/file_formats/parquet.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    if compression:
+        file_suffix = f'.{suffix}.gz'
+        parquet_compression='gzip'
+    else:
+        file_suffix = f'.{suffix}'
+        parquet_compression = None
+
+    filename = mkstemp(suffix=file_suffix, prefix=prefix, dir=dest_dir)[1]
+
+    dataframe = records_to_dataframe(records, schema, data_flattening_max_level)
```
[the remark this answers is not in the document]
Comment (koszti) [reply target_snowflake/file_formats/parquet.py:119]: pandas is using `pyarrow` to generate parquet files:

1. Pandas is installed by `snowflake-python-connector[pandas]`[here](https://github.com/transferwise/pipelinewise-target-snowflake/blob/68f0fe031d36219e1fc73542c421ae37d7b647cb/setup.py#L22) which is installing pandas with pyarrow. [here](https://github.com/snowflakedb/snowflake-connector-python/blob/a077643ed193fce94a23bf970cec35d938781712/setup.py#L42)
2. Pandas selecting Pyarrow automatically as a parquet generation engine when it's available. Docs [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_parquet.html)

The idea of generating parquet files based on the official snowflake-python-connector [write_pandas function](https://github.com/snowflakedb/snowflake-connector-python/blob/ed53074ace5a607471c62e1d83efe3075b9cd7c9/src/snowflake/connector/pandas_tools.py#L34). Usage of `to_parget` is [here](python/blob/ed53074ace5a607471c62e1d83efe3075b9cd7c9/src/snowflake/connector/pandas_tools.py#L126).

File: target_snowflake/upload_clients/s3_upload_client.py (hunk 1 above)
Comment (koszti) [reply target_snowflake/upload_clients/s3_upload_client.py:53]: The pattern is `<s3_key_prefix>pipelinewise_<stream>-<timestamp>-batch-<unique_temp_file_id>.<file_format>[.<compression>`

With real examples, with all the 4 different file types:
```
snowflake-imports-test/pipelinewise_performance_test_20210309-101648-516712_batch_762pt8p5.parquet.gz
snowflake-imports-test/pipelinewise_performance_test_20210309-101814-465401_batch_jrl2tdl8.parquet
snowflake-imports-test/pipelinewise_performance_test_20210309-101855-026878_batch_keu6f7w8.csv.gz
snowflake-imports-test/pipelinewise_performance_test_20210309-102006-808052_batch_1qz9tbmn.csv
```

```

File: target_snowflake/file_formats/parquet.py (hunk 2 above)
Comment (koszti) [reply target_snowflake/file_formats/parquet.py:23]: Selecting from parquet files requires the column names to be case-sensitive. The currently available `name` key in the dict is an sql friendly quoted and uppercased version of the column, which is fine for CSV but not compatible with selecting data from Parquet files:

**To Select from Parquet:**
```
SELECT ($1:col_1) COL_1, ($1:colTwo) COL_2, parse_json($1:col_3) COL_3 
  FROM '@foo_stage/foo_s3_key.csv' 
  (FILE_FORMAT => 'parquet_file_format')
  ;
```
Parquet files are a one column JSON representation of each record where element names are case-sensitive. We need to select the columns by `$1:caseSensitiveColumnName`. More info [here](https://docs.snowflake.com/en/user-guide/querying-semistructured.html#traversing-semi-structured-data). The case sensive version of the column name is the new `json_element_name` key in the dict.

**To Select from CSV:**
```
SELECT ($1) COL_1, ($2) COL_2, parse_json($3) COL_3
  FROM '@foo_stage/foo_s3_key.csv' 
  (FILE_FORMAT => 'csv_file_format)
```
CSV files are tabular and we can select columns directly by referring columns sequnce, by $1, $2, $n .

File: target_snowflake/db_sync.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Samira-El) [reply target_snowflake/db_sync.py:213]: Ok

File: target_snowflake/file_format.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Samira-El) [reply target_snowflake/file_format.py:52]: hmm ok

File: target_snowflake/upload_clients/s3_upload_client.py (hunk 1 above)
Comment (Samira-El) [reply target_snowflake/upload_clients/s3_upload_client.py:53]: Thanks.

File: target_snowflake/file_formats/parquet.py (hunk 2 above)
Comment (Samira-El) [reply target_snowflake/file_formats/parquet.py:23]: Very nice explanation, thanks! It's all clear.

File: target_snowflake/file_formats/csv.py [hunk 6]
```diff
@@ -3,11 +3,55 @@
...
+           "WHEN NOT MATCHED THEN " \
+           "INSERT ({}) " \
+           "VALUES ({})".format(
+        table_name,
+        ', '.join(["{}(${}) {}".format(c['trans'], i + 1, c['name']) for i, c in enumerate(columns)]),
+        stage_name,
+        s3_key,
+        file_format_name,
+        pk_merge_condition,
+        ', '.join(['{}=s.{}'.format(c['name'],
```
[the remark this answers is not in the document]
Comment (koszti) [reply target_snowflake/file_formats/csv.py]: addressed by 02bf2d042da6450bc14867513d5bf93865df680f

File: target_snowflake/file_format.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    """File Format class"""
+
+    def __init__(self, file_format: str, query_fn: Callable):
+        """Find the file format in Snowflake, detect its type and
+        initialise file format specific functions"""
+        # Detect file format type by querying it from Snowflake
+        self.file_format_type = self._detect_file_format_type(file_format, query_fn)
+
+        # Map file format specific functions dynamically
+        file_format_module = getattr(target_snowflake.file_formats, self.file_format_type)
```
[the remark this answers is not in the document]
Comment (koszti) [reply target_snowflake/file_format.py]: Addressed by 2cca9fee50f94979d4745dbb3547f1e639a1f797

**Review by Samira-El — APPROVED** [review]

**Pull request merged**